HeterogeneousCollections/DiffList.fsi

```diff
@@ -0,0 +1,58 @@
+namespace HCollections
+
+open TypeEquality
+
+/// DiffList is a heterogeneous list similar to HList, except that it has a "unification" variable for the base.
+/// This allows for an append function between two DiffLists.
+[<NoComparison>]
+[<NoEquality>]
+type DiffList<'ty, 'v>
+
+/// DiffListFolder allows you to perform a fold over an DiffList.
+/// The single type parameter, 'state, denotes the type of the value
+/// that you want the fold to return.
+type 'state DiffListFolder =
+
+    /// F takes the current state, the next element in the DiffList and returns a new state.
+    /// Because elements in the DiffList may have arbitrary type, F must be generic on
+    /// the element type, i.e. can be called for any element type.
+    abstract Folder<'a> : 'state -> 'a -> 'state
+
+module DiffList =
+
+    /// Congruence proof for DiffList - given a proof of equality between two types 'ty1 and 'ty2,
+    /// returns a proof that DiffList<'ty1, 'v> and DiffList<'ty2, 'v> are the same type.
+    val cong : Teq<'ty1, 'ty2> -> Teq<DiffList<'ty1, 'v>, DiffList<'ty2, 'v>>
+
+    /// Congruence proof for DiffList unification variable - given a proof of equality between two types 'u and 'v,
+    /// returns a proof that DiffList<'ty, 'u> and DiffList<'ty, 'v> are the same type.
+    val congUnification : Teq<'u, 'v> -> Teq<DiffList<'ty, 'u>, DiffList<'ty, 'v>>
```

Comment:
If we instead did this:

```fsharp
val cong : Teq<'elts1, 'elts2> -> Teq<'tailSlot1, 'tailSlot2> -> Teq<DiffList<_, _>, DiffList<_, _>>
```

then we wouldn't need two functions. That may or may not be good; it would avoid the question of naming for "unification".

By the way, could we please have explicit type variables `val cong<'ty1, ...>`? Teq-oriented programming is even more assisted than usual by the user annotating types, so it's good manners for us to let them specify the types explicitly.